File: competence-database/src/main/scala/uzuzjmd/competence/owl/dao/AbstractEvidenceLink.scala
package uzuzjmd.competence.owl.dao

import uzuzjmd.competence.owl.access.CompOntologyManager
import com.hp.hpl.jena.ontology.Individual
import uzuzjmd.competence.owl.ontology.CompOntClass
import uzuzjmd.competence.owl.ontology.CompObjectProperties
import uzuzjmd.competence.owl.access.CompOntologyAccess
import uzuzjmd.competence.owl.access.CompOntologyAccessScala
//import com.google.gwt.thirdparty.guava.common.collect.Collections2.PermutationCollection

case class AbstractEvidenceLink(

    val comp: CompOntologyManager,
    val identifier2: String = null,
    val creator: User,
    val linkedUser: User,
    val courseContexts: CourseContext,
    val evidenceActivity: EvidenceActivity = null,
    val dateCreated: java.lang.Long,
    val isValidated: java.lang.Boolean = false,
    val competence: Competence = null,
    val comments: List[Comment] = null) extends CompetenceOntologyDao(comp, CompOntClass.AbstractEvidenceLink, CompOntologyAccessScala.computeEncodedStringForLink(identifier2, competence, evidenceActivity)) {

  def CREATED = "datecreated"
  def ISVALIDATED = "isValidated"

  def this(comp: CompOntologyManager, identifier2: String, competence: Competence = null, evidenceActivity: EvidenceActivity = null) = this(comp, CompOntologyAccessScala.computeEncodedStringForLink(identifier2, competence, evidenceActivity), null, null, null, null, null, null, competence, null)

  def this(comp: CompOntologyManager, identifier2: String, competence2: Competence, evidenceActivity2: EvidenceActivity, comments: List[Comment]) = this(comp, CompOntologyAccessScala.computeEncodedStringForLink(identifier2, competence2, evidenceActivity2), null, null, null, null, null, null, competence2, comments)

  //  def computeEncodedString: String = {
  //    if (identifier2 != null) {
  //      return identifier2
  //    } else {
  //      if (competence == null || evidenceActivity == null) {
  //        throw new Exception("evidenceActivity and competence need to be defined")
  //      }
  //      return competence.getId + evidenceActivity.getId
  //    }
  //  }

  @Override
  protected def deleteMore() {
    if (comments != null) {
      comments.foreach(x => x.delete)
    }
  }

  @Override
  protected def persistMore() {
    if (comments != null) {
      linkComments(comments)
    }
    linkedUser.persist
    createEdgeWith(linkedUser, CompObjectProperties.UserOfLink)
    creator.persist
    createEdgeWith(CompObjectProperties.createdBy, creator)
    courseContexts.persist
    createEdgeWith(CompObjectProperties.LinkOfCourseContext, courseContexts)
    evidenceActivity.persist
    createEdgeWith(evidenceActivity, CompObjectProperties.ActivityOf)
    competence.persist(false)
    createEdgeWith(CompObjectProperties.linksCompetence, competence)
    addDataField(CREATED, dateCreated)
    addDataField(ISVALIDATED, isValidated)
  }

  def linkComments(comments: List[Comment]) {
    comments.foreach(linkComment)
  }

  def linkComment(comment: Comment) {
    comment.persist
    createEdgeWith(comment, CompObjectProperties.CommentOf)
  }

  @Override
  def getFullDao(): AbstractEvidenceLink = {
    val comments = getAssociatedStandardDaosAsDomain(CompObjectProperties.CommentOf, classOf[Comment])
    val creator = getAssociatedStandardDaosAsRange(CompObjectProperties.createdBy, classOf[User]).head
    val courseContext = getAssociatedStandardDaosAsRange(CompObjectProperties.LinkOfCourseContext, classOf[CourseContext])
    val evidenceActivity = getAssociatedStandardDaosAsDomain(CompObjectProperties.ActivityOf, classOf[EvidenceActivity])
    return new AbstractEvidenceLink(comp, identifier, creator.getFullDao, null, null, null, getDataFieldLong(CREATED), getDataFieldBoolean(ISVALIDATED), null, comments)
  }

  def getAllLinkedUsers(): List[User] = {
    return getAssociatedStandardDaosAsDomain(CompObjectProperties.UserOfLink, classOf[User]).map(x => x.getFullDao.asInstanceOf[User])
  }

  def getAllActivities(): List[EvidenceActivity] = {
    return getAssociatedStandardDaosAsDomain(CompObjectProperties.ActivityOf, classOf[EvidenceActivity]).map(x => x.getFullDao.asInstanceOf[EvidenceActivity])
  }

  def getAllCourseContexts(): List[CourseContext] = {
    return getAssociatedStandardDaosAsRange(CompObjectProperties.LinkOfCourseContext, classOf[CourseContext]).map(x => x.getFullDao.asInstanceOf[CourseContext])
  }

  def getAllLinkedCompetences(): List[Competence] = {
    return getAssociatedSingletonDaosAsRange(CompObjectProperties.linksCompetence, classOf[Competence]).map(x => x.getFullDao.asInstanceOf[Competence])
  }

}

File: competence-database/src/main/scala/uzuzjmd/competence/monopersistence/daos/AbstractEvidenceLink.java
package uzuzjmd.competence.monopersistence.daos;

import uzuzjmd.competence.persistence.ontology.CompObjectProperties;

import java.util.List;

/**
 * Created by dehne on 11.01.2016.
 */
public class AbstractEvidenceLink extends DaoAbstractImpl implements Cascadable {

    public User creator;
    public User linkedUser;
    public CourseContext courseContext;
    public EvidenceActivity evidenceActivity;
    public Long dateCreated;
    public Boolean isValidated;
    public Competence competence;
    public List<Comment> comments;

    public AbstractEvidenceLink(String id) {
        super(id);
    }

    public AbstractEvidenceLink(User creator,
                                User linkedUser,
                                CourseContext courseContext,
                                EvidenceActivity evidenceActivity,
                                Long dateCreated,
                                Boolean isValidated,
                                Competence competence,
                                List<Comment> comments) {
        super(competence.getId() + evidenceActivity.getId());
        this.creator = creator;
        this.linkedUser = linkedUser;
        this.courseContext = courseContext;
        this.evidenceActivity = evidenceActivity;
        this.dateCreated = dateCreated;
        this.isValidated = isValidated;
        this.competence = competence;
        this.comments = comments;
    }



    public List<Comment> getComments() {
        return comments;
    }

    public void setComments(List<Comment> comments) {
        this.comments = comments;
    }

    public Competence getCompetence() {
        return competence;
    }

    public void setCompetence(Competence competence) {
        this.competence = competence;
    }

    public Boolean getValidated() {
        return isValidated;
    }

    public void setValidated(Boolean validated) {
        isValidated = validated;
    }

    public Long getDateCreated() {
        return dateCreated;
    }

    public void setDateCreated(Long dateCreated) {
        this.dateCreated = dateCreated;
    }

    public EvidenceActivity getEvidenceActivity() {
        return evidenceActivity;
    }

    public void setEvidenceActivity(EvidenceActivity evidenceActivity) {
        this.evidenceActivity = evidenceActivity;
    }

    public CourseContext getCourseContext() {
        return courseContext;
    }

    public void setCourseContext(CourseContext courseContext) {
        this.courseContext = courseContext;
    }

    public User getLinkedUser() {
        return linkedUser;
    }

    public void setLinkedUser(User linkedUser) {
        this.linkedUser = linkedUser;
    }

    public User getCreator() {
        return creator;
    }

    public void setCreator(User creator) {
        this.creator = creator;
    }

    @Override
    public void persistMore() throws Exception{
        linkedUser.persist();
        createEdgeWith(linkedUser, CompObjectProperties.UserOfLink);
        creator.persist();
        createEdgeWith(CompObjectProperties.createdBy, creator);
        courseContext.persist();
        createEdgeWith(CompObjectProperties.LinkOfCourseContext, courseContext);
        evidenceActivity.persist();
        createEdgeWith(evidenceActivity, CompObjectProperties.ActivityOf);
        competence.persist();
        createEdgeWith(CompObjectProperties.linksCompetence, competence);
        this.persist();
    }


    public void linkComments(List<Comment> comments) throws Exception {
        for (Comment comment : comments) {
            linkComment(comment);
        }
    }

    public void linkComment(Comment comment) throws Exception {
            comment.persist();
            createEdgeWith(comment, CompObjectProperties.CommentOf);
    }

    public List<User> getAllLinkedUsers() throws Exception {
            return getAssociatedDaosAsDomain(CompObjectProperties.UserOfLink, User.class);
    }

    public List<EvidenceActivity> getAllActivities() throws Exception {
            return getAssociatedDaosAsDomain(CompObjectProperties.ActivityOf, EvidenceActivity.class);
    }

    public List<CourseContext> getAllCourseContexts() throws Exception {
            return getAssociatedDaosAsRange(CompObjectProperties.LinkOfCourseContext, CourseContext.class);
    }

    public List<Competence> getAllLinkedCompetences() throws Exception {
        return  getAssociatedDaosAsRange(CompObjectProperties.linksCompetence, Competence.class);
    }


}
